bindings/pyroot/test/rdataframe_asnumpy.py

```diff
@@ -0,0 +1,167 @@
+        col_names = ["col_{}".format(d) for d in dtypes]
+
+        col_vars = []
+        for dtype in dtypes:
+            if "F" in dtype:
+                var = np.empty(1, dtype=np.float32)
+            elif "D" in dtype:
+                var = np.empty(1, dtype=np.float64)
+            elif "I" in dtype:
+                var = np.empty(1, dtype=np.int32)
+            elif "i" in dtype:
+                var = np.empty(1, dtype=np.uint32)
+            elif "L" in dtype:
+                var = np.empty(1, dtype=np.int64)
+            elif "l" in dtype:
+                var = np.empty(1, dtype=np.uint64)
+            elif "S" in dtype:
+                var = np.empty(1, dtype=np.int16)
+            elif "s" in dtype:
+                var = np.empty(1, dtype=np.uint16)
+            elif "B" in dtype:
+                var = np.empty(1, dtype=np.int8)
+            elif "b" in dtype:
+                var = np.empty(1, dtype=np.uint8)
+            elif "O" in dtype:
+                var = np.empty(1, dtype=np.uint8)
+            else:
+                raise Exception(
+                    "Type {} not known to create branch.".format(dtype))
+            col_vars.append(var)
+
+        for dtype, name, var in zip(dtypes, col_names, col_vars):
+            tree.Branch(name, var, name + "/" + dtype)
+
+        reference = {col: [] for col in col_names}
+        for i in range(5):
+            for i_var, var in enumerate(col_vars):
+                var[0] = i
+                reference[col_names[i_var]].append(var[0])
+            tree.Fill()
+        reference = {col: np.array(reference[col]) for col in reference}
+
+        return tree, reference, dtypes, col_names, col_vars
+
+    def test_branch_dtypes(self):
+        root_dtypes = ["S", "s", "I", "i", "L", "l", "F", "D"]
+        tree, ref, _, col_names, _ = self.make_tree(*root_dtypes)
+        df = ROOT.RDataFrame(tree)
+        npy = df.AsNumpy()
+        for col in col_names:
+            self.assertTrue(all(npy[col] == ref[col]))
+
+    def test_read_array(self):
+        ROOT.gInterpreter.Declare("""
+        std::array<unsigned int, 3> create_array(unsigned int n) {
+            return std::array<unsigned int, 3>({n, n, n});
+        }
+        """)
+        df = ROOT.RDataFrame(5).Define("x", "create_array(tdfentry_)")
+        npy = df.AsNumpy()
+        self.assertEqual(npy["x"].size, 5)
+        self.assertEqual(list(npy["x"][0]), [0, 0, 0])
+        self.assertIn("ROOT.array<unsigned int,3>", str(type(npy["x"][0])))
+
+    def test_read_th1f(self):
+        ROOT.gInterpreter.Declare("""
+        TH1F create_histo(unsigned int n) {
+            const auto str = TString::Format("h%i", n);
+            return TH1F(str, str, 4, 0, 1);
+        }
+        """)
+        df = ROOT.RDataFrame(5).Define("x", "create_histo(tdfentry_)")
+        npy = df.AsNumpy()
+        self.assertEqual(npy["x"].size, 5)
+        self.assertIn("ROOT.TH1F", str(type(npy["x"][0])))
```

Comment:
Why `assertIn` and not `assertEqual` here?

Reply:
the exact returned string is `<class 'ROOT.TH1F'>`. i was a little bit lazy here!